File: apps/agent/routers/canvas.py
from fastapi import APIRouter, Request, HTTPException
#from routers.agent import chat
from services.chat_service import handle_chat
from services.api_client_service import api_client_service
import asyncio
import json

router = APIRouter(prefix="/api/canvas")

@router.get("/list")
async def list_canvases():
    try:
        result = await api_client_service.list_canvases()
        return result
    except Exception as e:
        print(f"API client failed: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch canvases from server")

@router.post("/create")
async def create_canvas(request: Request):
    data = await request.json()
    id = data.get('canvas_id')
    name = data.get('name')

    asyncio.create_task(handle_chat(data))
    try:
        await api_client_service.create_canvas(name, id)
        return {"id": id }
    except Exception as e:
        print(f"API client failed: {e}")
        raise HTTPException(status_code=500, detail="Failed to create canvas")

@router.get("/{id}")
async def get_canvas(id: str):
    try:
        result = await api_client_service.get_canvas_data(id)
        if result is not None:
            return result
        else:
            raise HTTPException(status_code=404, detail="Canvas not found")
    except Exception as e:
        print(f"API client failed: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch canvas data")

@router.post("/{id}/save")
async def save_canvas(id: str, request: Request):
    payload = await request.json()
    # 优先使用API客户端，如果失败则回退到本地数据库
    try:
        await api_client_service.save_canvas_data(id, payload['data'], payload.get('thumbnail'))
    except Exception as e:
        print(f"API client failed, falling back to local DB: {e}")
        # 注释掉本地数据库回退，因为已经迁移到API
        # data_str = json.dumps(payload['data'])
        # await db_service.save_canvas_data(id, data_str, payload.get('thumbnail'))
    return {"id": id }

@router.put("/{id}/timeline/assets/starttime")
async def update_timeline_asset_start_times(id: str, request: Request):
    """更新Timeline资产的startTime"""
    payload = await request.json()
    try:
        result = await api_client_service.update_timeline_asset_start_times(id, payload['assets'])
        return result
    except Exception as e:
        print(f"Failed to update timeline asset startTimes: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/{id}/rename")
async def rename_canvas(id: str, request: Request):
    data = await request.json()
    name = data.get('name')
    try:
        await api_client_service.rename_canvas(id, name)
        return {"id": id }
    except Exception as e:
        print(f"API client failed: {e}")
        raise HTTPException(status_code=500, detail="Failed to rename canvas")

@router.delete("/{id}/delete")
async def delete_canvas(id: str):
    try:
        await api_client_service.delete_canvas(id)
        return {"id": id }
    except Exception as e:
        print(f"API client failed: {e}")
        raise HTTPException(status_code=500, detail="Failed to delete canvas")

File: apps/web/lib/nolanx/api/canvas.ts
import { CanvasData, Message, Session } from '../types/types'
import { apiClient, ApiResponse } from '@/lib/api/client'

export type CanvasItem = {
  id: string
  name: string
  description?: string
  thumbnail?: string
  data?: any
  created_at: string
}

export type ListCanvasesResponse = {
  canvases: CanvasItem[];
  total: number;
  page: number;
  limit: number;
}

export interface CreateCanvasParams {
  canvas_id?: string
  name?: string
  messages: Message[]
  preferred_language?: string
}

export interface SaveCanvasParams {
  data: CanvasData
  thumbnail: string
}

export interface RegenerateTimelineAssetParams {
  prompt?: string
}

// 获取画布列表
export const listCanvases = async (): Promise<CanvasItem[]> => {
  const { data: res } = await apiClient.get<ApiResponse<ListCanvasesResponse>>('/canvas');
  return res.canvases;
};

// 创建画布
export const createCanvas = async (data: CreateCanvasParams): Promise<{ canvas_id: string }> => {
  const { data: res } = await apiClient.post<ApiResponse<{ canvas_id: string }>>('/canvas/create', data);
  return res;
}

// 获取画布详情
export const getCanvas = async (id: string): Promise<{ data: CanvasData; name: string; sessions: Session[] }> => {
  const { data: res } = await apiClient.get<ApiResponse<{ data: CanvasData; name: string; sessions: Session[] }>>(`/canvas/${id}`);
  return res;
}

// 保存画布
export const saveCanvas = async (id: string, payload: SaveCanvasParams): Promise<void> => {
  const { data: res } = await apiClient.post<ApiResponse<void>>(`/canvas/${id}/save`, payload);
  return res;
}

// 重命名画布
export const renameCanvas = async (id: string, name: string): Promise<void> => {
  const { data: res } = await apiClient.post<ApiResponse<void>>(`/canvas/${id}/rename`, { name });
  return res;
}

// 删除画布
export const deleteCanvas = async (id: string): Promise<void> => {
  const { data: res } = await apiClient.post<ApiResponse<void>>(`/canvas/${id}/delete`, {});
  return res;
}

// 更新Timeline资产的startTime
export const updateTimelineAssetStartTimes = async (
  id: string,
  assets: Array<{ id: string; startTime: number }>
): Promise<{ success: boolean; updatedCount: number }> => {
  const { data: res } = await apiClient.post<ApiResponse<{ success: boolean; updatedCount: number }>>(
    `/canvas/${id}/timeline/assets/starttime`,
    { assets }
  );
  return res;
}

export const regenerateTimelineAsset = async (
  id: string,
  assetId: string,
  payload: RegenerateTimelineAssetParams = {}
): Promise<{ success: boolean; asset: any; credits_consumed: number }> => {
  const { data: res } = await apiClient.post<ApiResponse<{ success: boolean; asset: any; credits_consumed: number }>>(
    `/canvas/${id}/timeline/asset/${assetId}/regenerate`,
    payload,
    {
      timeout: 10 * 60 * 1000,
    }
  );
  return res;
}

// 获取分享的画布详情（无需认证）
export const getSharedCanvas = async (id: string): Promise<{ id: string; name: string; canvas_id: string; data: CanvasData; thumbnail?: string; created_at: string; updated_at: string; sessions: Session[] }> => {
  const { data: res } = await apiClient.get<ApiResponse<{ id: string; name: string; canvas_id: string; data: CanvasData; thumbnail?: string; created_at: string; updated_at: string; sessions: Session[] }>>(`/canvas/share/${id}`);
  return res;
}

// 获取分享的聊天历史（无需认证）
export const getSharedChatHistory = async (sessionId: string): Promise<Message[]> => {
  const { data: res } = await apiClient.get<ApiResponse<any>>(`/canvas/share/session/msgs/${sessionId}`);
  const list = res.map(({ content }: any) => content) as Message[]
  return list;
}
